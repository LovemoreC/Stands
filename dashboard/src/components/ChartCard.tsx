import React from 'react';
import {
  Bar,
  BarChart,
  CartesianGrid,
  Cell,
  Legend,
  Line,
  LineChart,
  Pie,
  PieChart,
  ResponsiveContainer,
  Tooltip,
  XAxis,
  YAxis,
} from 'recharts';

export interface ChartDatum {
  label: string;
  value: number;
}

interface BaseChartCardProps {
  title: string;
  description?: string;
  data: ChartDatum[];
  ariaLabel?: string;
  chartHeight?: number;
  valueLabel?: string;
}

interface ChartRenderContext {
  data: ChartDatum[];
  colors: string[];
  valueLabel: string;
}

type ChartRenderer = (ctx: ChartRenderContext) => React.ReactNode;

const COLORS = ['#4C51BF', '#48BB78', '#ED8936', '#3182CE', '#9F7AEA', '#F56565'];

const ChartCard: React.FC<BaseChartCardProps & { renderChart: ChartRenderer }> = ({
  title,
  description,
  data,
  ariaLabel,
  chartHeight = 260,
  valueLabel = 'Value',
  renderChart,
}) => {
  const containerRef = React.useRef<HTMLDivElement | null>(null);
  const [width, setWidth] = React.useState(400);

  React.useEffect(() => {
    const updateSize = () => {
      const nextWidth = containerRef.current?.clientWidth;
      if (nextWidth && nextWidth !== width) {
        setWidth(nextWidth);
      }
    };
    updateSize();
    window.addEventListener('resize', updateSize);
    return () => {
      window.removeEventListener('resize', updateSize);
    };
  }, [width]);

  const chartAria = ariaLabel ?? `${title} visualisation`;

  return (
    <section className="chart-card">
      <header className="chart-card__header">
        <h3>{title}</h3>
        {description && <p className="chart-card__description">{description}</p>}
      </header>
      <div className="chart-card__body">
        <div
          className="chart-card__visual"
          role="img"
          aria-label={chartAria}
          ref={containerRef}
          style={{ width: '100%' }}
        >
          {data.length > 0 ? (
            <ResponsiveContainer width={width} height={chartHeight}>
              {renderChart({ data, colors: COLORS, valueLabel })}
            </ResponsiveContainer>
          ) : (
            <p className="chart-card__empty" role="status">
              No data available
            </p>
          )}
        </div>
        {data.length > 0 && (
          <table className="chart-card__table">
            <caption className="visually-hidden">{`${title} data table`}</caption>
            <thead>
              <tr>
                <th scope="col">Label</th>
                <th scope="col">{valueLabel}</th>
              </tr>
            </thead>
            <tbody>
              {data.map((datum) => (
                <tr key={datum.label}>
                  <td>{datum.label}</td>
                  <td>{datum.value.toLocaleString()}</td>
                </tr>
              ))}
            </tbody>
          </table>
        )}
      </div>
    </section>
  );
};

export const BarChartCard: React.FC<BaseChartCardProps> = (props) => (
  <ChartCard
    {...props}
    renderChart={({ data, colors, valueLabel }) => (
      <BarChart data={data}>
        <CartesianGrid strokeDasharray="3 3" />
        <XAxis dataKey="label" interval={0} angle={data.length > 3 ? -20 : 0} textAnchor={data.length > 3 ? 'end' : 'middle'} height={data.length > 3 ? 80 : undefined} />
        <YAxis allowDecimals={false} />
        <Tooltip formatter={(value: number) => value.toLocaleString()} labelFormatter={(label) => `${label}`} />
        <Legend />
        <Bar dataKey="value" name={valueLabel}>
          {data.map((entry, index) => (
            <Cell key={entry.label} fill={colors[index % colors.length]} />
          ))}
        </Bar>
      </BarChart>
    )}
  />
);

export const LineChartCard: React.FC<BaseChartCardProps> = (props) => (
  <ChartCard
    {...props}
    renderChart={({ data, colors, valueLabel }) => (
      <LineChart data={data}>
        <CartesianGrid strokeDasharray="3 3" />
        <XAxis dataKey="label" interval={0} angle={data.length > 4 ? -20 : 0} textAnchor={data.length > 4 ? 'end' : 'middle'} height={data.length > 4 ? 80 : undefined} />
        <YAxis allowDecimals={false} />
        <Tooltip formatter={(value: number) => value.toLocaleString()} labelFormatter={(label) => `${label}`} />
        <Legend />
        <Line type="monotone" dataKey="value" name={valueLabel} stroke={colors[0]} strokeWidth={3} dot={{ r: 4 }} />
      </LineChart>
    )}
  />
);

export const PieChartCard: React.FC<BaseChartCardProps> = (props) => (
  <ChartCard
    {...props}
    renderChart={({ data, colors, valueLabel }) => (
      <PieChart>
        <Tooltip formatter={(value: number, _name, payload) => [value.toLocaleString(), payload?.payload?.label]} />
        <Legend />
        <Pie data={data} dataKey="value" nameKey="label" label={({ name, value }) => `${name}: ${value}`}> 
          {data.map((entry, index) => (
            <Cell key={entry.label} fill={colors[index % colors.length]} />
          ))}
        </Pie>
      </PieChart>
    )}
  />
);

export const toChartData = (record: Record<string, number> | null | undefined): ChartDatum[] => {
  if (!record) return [];
  return Object.entries(record)
    .filter(([, value]) => typeof value === 'number' && !Number.isNaN(value))
    .map(([label, value]) => ({ label: formatLabel(label), value }));
};

const formatLabel = (label: string) =>
  label
    .replace(/[_-]+/g, ' ')
    .replace(/\b\w/g, (char) => char.toUpperCase());

export default ChartCard;

