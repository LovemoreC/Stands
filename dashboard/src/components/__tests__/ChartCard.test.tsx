import React from 'react';
import { render, screen } from '@testing-library/react';
import { describe, expect, it } from 'vitest';
import { BarChartCard, LineChartCard, PieChartCard, toChartData } from '../ChartCard';

describe('shared chart components', () => {
  const sampleData = [
    { label: 'Available', value: 12 },
    { label: 'Reserved', value: 5 },
    { label: 'Sold', value: 7 },
  ];

  it('renders a bar chart with a data table fallback', () => {
    const { asFragment } = render(
      <BarChartCard
        title="Inventory overview"
        description="Distribution of inventory"
        data={sampleData}
        valueLabel="Properties"
      />,
    );

    expect(screen.getByRole('img', { name: /inventory overview/i })).toBeInTheDocument();
    expect(screen.getByRole('table')).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders a line chart', () => {
    render(
      <LineChartCard
        title="Deposit performance"
        description="Weekly deposits"
        data={sampleData}
        valueLabel="Deposits"
      />,
    );

    expect(screen.getByRole('img', { name: /deposit performance/i })).toBeInTheDocument();
  });

  it('renders a pie chart', () => {
    render(
      <PieChartCard
        title="Mandate statuses"
        description="Breakdown of mandate statuses"
        data={sampleData}
        valueLabel="Mandates"
      />,
    );

    expect(screen.getByRole('img', { name: /mandate statuses/i })).toBeInTheDocument();
  });

  it('converts records to chart data', () => {
    const chartData = toChartData({ pending_review: 3, approved: 4 });
    expect(chartData).toEqual([
      { label: 'Pending Review', value: 3 },
      { label: 'Approved', value: 4 },
    ]);
  });
});

