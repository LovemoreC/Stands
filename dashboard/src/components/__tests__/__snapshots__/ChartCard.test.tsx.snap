// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`shared chart components > renders a bar chart with a data table fallback 1`] = `
<DocumentFragment>
  <section
    class="chart-card"
  >
    <header
      class="chart-card__header"
    >
      <h3>
        Inventory overview
      </h3>
      <p
        class="chart-card__description"
      >
        Distribution of inventory
      </p>
    </header>
    <div
      class="chart-card__body"
    >
      <div
        aria-label="Inventory overview visualisation"
        class="chart-card__visual"
        role="img"
        style="width: 100%;"
      >
        <div
          class="recharts-responsive-container"
          style="width: 400px; height: 260px; min-width: 0;"
        >
          <div
            style="width: 0px; height: 0px; overflow: visible;"
          >
            <div
              class="recharts-wrapper"
              style="position: relative; cursor: default; width: 400px; height: 260px;"
            >
              <div
                class="recharts-tooltip-wrapper"
                style="visibility: hidden; pointer-events: none; position: absolute; top: 0px; left: 0px;"
                tabindex="-1"
                xmlns="http://www.w3.org/1999/xhtml"
              >
                <div
                  aria-live="assertive"
                  class="recharts-default-tooltip"
                  role="status"
                  style="margin: 0px; padding: 10px; background-color: rgb(255, 255, 255); white-space: nowrap;"
                >
                  <p
                    class="recharts-tooltip-label"
                    style="margin: 0px;"
                  />
                </div>
              </div>
              <div
                class="recharts-legend-wrapper"
                style="position: absolute; width: 390px; height: auto; left: 5px; bottom: 5px;"
              >
                <ul
                  class="recharts-default-legend"
                  style="padding: 0px; margin: 0px; text-align: center;"
                >
                  <li
                    class="recharts-legend-item legend-item-0"
                    style="display: inline-block; margin-right: 10px;"
                  >
                    <svg
                      aria-label="Properties legend icon"
                      class="recharts-surface"
                      height="14"
                      style="display: inline-block; vertical-align: middle; margin-right: 4px;"
                      viewBox="0 0 32 32"
                      width="14"
                    >
                      <title />
                      <desc />
                      <path
                        class="recharts-legend-icon"
                        d="M0,4h32v24h-32z"
                        stroke="none"
                      />
                    </svg>
                    <span
                      class="recharts-legend-item-text"
                    >
                      Properties
                    </span>
                  </li>
                </ul>
              </div>
              <svg
                class="recharts-surface"
                height="260"
                role="application"
                style="width: 100%; height: 100%;"
                tabindex="0"
                viewBox="0 0 400 260"
                width="400"
              >
                <title />
                <desc />
                <defs>
                  <clippath
                    id="recharts1-clip"
                  >
                    <rect
                      height="220"
                      width="330"
                      x="65"
                      y="5"
                    />
                  </clippath>
                </defs>
                <g
                  class="recharts-cartesian-grid"
                >
                  <g
                    class="recharts-cartesian-grid-horizontal"
                  >
                    <line
                      fill="none"
                      height="220"
                      stroke="#ccc"
                      stroke-dasharray="3 3"
                      width="330"
                      x="65"
                      x1="65"
                      x2="395"
                      y="5"
                      y1="225"
                      y2="225"
                    />
                    <line
                      fill="none"
                      height="220"
                      stroke="#ccc"
                      stroke-dasharray="3 3"
                      width="330"
                      x="65"
                      x1="65"
                      x2="395"
                      y="5"
                      y1="170"
                      y2="170"
                    />
                    <line
                      fill="none"
                      height="220"
                      stroke="#ccc"
                      stroke-dasharray="3 3"
                      width="330"
                      x="65"
                      x1="65"
                      x2="395"
                      y="5"
                      y1="115"
                      y2="115"
                    />
                    <line
                      fill="none"
                      height="220"
                      stroke="#ccc"
                      stroke-dasharray="3 3"
                      width="330"
                      x="65"
                      x1="65"
                      x2="395"
                      y="5"
                      y1="60"
                      y2="60"
                    />
                    <line
                      fill="none"
                      height="220"
                      stroke="#ccc"
                      stroke-dasharray="3 3"
                      width="330"
                      x="65"
                      x1="65"
                      x2="395"
                      y="5"
                      y1="5"
                      y2="5"
                    />
                  </g>
                  <g
                    class="recharts-cartesian-grid-vertical"
                  >
                    <line
                      fill="none"
                      height="220"
                      stroke="#ccc"
                      stroke-dasharray="3 3"
                      width="330"
                      x="65"
                      x1="120"
                      x2="120"
                      y="5"
                      y1="5"
                      y2="225"
                    />
                    <line
                      fill="none"
                      height="220"
                      stroke="#ccc"
                      stroke-dasharray="3 3"
                      width="330"
                      x="65"
                      x1="230"
                      x2="230"
                      y="5"
                      y1="5"
                      y2="225"
                    />
                    <line
                      fill="none"
                      height="220"
                      stroke="#ccc"
                      stroke-dasharray="3 3"
                      width="330"
                      x="65"
                      x1="340"
                      x2="340"
                      y="5"
                      y1="5"
                      y2="225"
                    />
                    <line
                      fill="none"
                      height="220"
                      stroke="#ccc"
                      stroke-dasharray="3 3"
                      width="330"
                      x="65"
                      x1="65"
                      x2="65"
                      y="5"
                      y1="5"
                      y2="225"
                    />
                    <line
                      fill="none"
                      height="220"
                      stroke="#ccc"
                      stroke-dasharray="3 3"
                      width="330"
                      x="65"
                      x1="395"
                      x2="395"
                      y="5"
                      y1="5"
                      y2="225"
                    />
                  </g>
                </g>
                <g
                  class="recharts-layer recharts-cartesian-axis recharts-xAxis xAxis"
                >
                  <line
                    angle="0"
                    class="recharts-cartesian-axis-line"
                    fill="none"
                    height="30"
                    orientation="bottom"
                    stroke="#666"
                    text-anchor="middle"
                    width="330"
                    x="65"
                    x1="65"
                    x2="395"
                    y="225"
                    y1="225"
                    y2="225"
                  />
                  <g
                    class="recharts-cartesian-axis-ticks"
                  >
                    <g
                      class="recharts-layer recharts-cartesian-axis-tick"
                    >
                      <line
                        angle="0"
                        class="recharts-cartesian-axis-tick-line"
                        fill="none"
                        height="30"
                        orientation="bottom"
                        stroke="#666"
                        text-anchor="middle"
                        width="330"
                        x="65"
                        x1="120"
                        x2="120"
                        y="225"
                        y1="231"
                        y2="225"
                      />
                      <text
                        class="recharts-text recharts-cartesian-axis-tick-value"
                        fill="#666"
                        height="30"
                        orientation="bottom"
                        stroke="none"
                        text-anchor="middle"
                        width="330"
                        x="120"
                        y="233"
                      >
                        <tspan
                          dy="0.71em"
                          x="120"
                        >
                          Available
                        </tspan>
                      </text>
                    </g>
                    <g
                      class="recharts-layer recharts-cartesian-axis-tick"
                    >
                      <line
                        angle="0"
                        class="recharts-cartesian-axis-tick-line"
                        fill="none"
                        height="30"
                        orientation="bottom"
                        stroke="#666"
                        text-anchor="middle"
                        width="330"
                        x="65"
                        x1="230"
                        x2="230"
                        y="225"
                        y1="231"
                        y2="225"
                      />
                      <text
                        class="recharts-text recharts-cartesian-axis-tick-value"
                        fill="#666"
                        height="30"
                        orientation="bottom"
                        stroke="none"
                        text-anchor="middle"
                        width="330"
                        x="230"
                        y="233"
                      >
                        <tspan
                          dy="0.71em"
                          x="230"
                        >
                          Reserved
                        </tspan>
                      </text>
                    </g>
                    <g
                      class="recharts-layer recharts-cartesian-axis-tick"
                    >
                      <line
                        angle="0"
                        class="recharts-cartesian-axis-tick-line"
                        fill="none"
                        height="30"
                        orientation="bottom"
                        stroke="#666"
                        text-anchor="middle"
                        width="330"
                        x="65"
                        x1="340"
                        x2="340"
                        y="225"
                        y1="231"
                        y2="225"
                      />
                      <text
                        class="recharts-text recharts-cartesian-axis-tick-value"
                        fill="#666"
                        height="30"
                        orientation="bottom"
                        stroke="none"
                        text-anchor="middle"
                        width="330"
                        x="340"
                        y="233"
                      >
                        <tspan
                          dy="0.71em"
                          x="340"
                        >
                          Sold
                        </tspan>
                      </text>
                    </g>
                  </g>
                </g>
                <g
                  class="recharts-layer recharts-cartesian-axis recharts-yAxis yAxis"
                >
                  <line
                    angle="0"
                    class="recharts-cartesian-axis-line"
                    fill="none"
                    height="220"
                    orientation="left"
                    stroke="#666"
                    width="60"
                    x="5"
                    x1="65"
                    x2="65"
                    y="5"
                    y1="5"
                    y2="225"
                  />
                  <g
                    class="recharts-cartesian-axis-ticks"
                  >
                    <g
                      class="recharts-layer recharts-cartesian-axis-tick"
                    >
                      <line
                        angle="0"
                        class="recharts-cartesian-axis-tick-line"
                        fill="none"
                        height="220"
                        orientation="left"
                        stroke="#666"
                        width="60"
                        x="5"
                        x1="59"
                        x2="65"
                        y="5"
                        y1="225"
                        y2="225"
                      />
                      <text
                        class="recharts-text recharts-cartesian-axis-tick-value"
                        fill="#666"
                        height="220"
                        orientation="left"
                        stroke="none"
                        text-anchor="end"
                        width="60"
                        x="57"
                        y="225"
                      >
                        <tspan
                          dy="0.355em"
                          x="57"
                        >
                          0
                        </tspan>
                      </text>
                    </g>
                    <g
                      class="recharts-layer recharts-cartesian-axis-tick"
                    >
                      <line
                        angle="0"
                        class="recharts-cartesian-axis-tick-line"
                        fill="none"
                        height="220"
                        orientation="left"
                        stroke="#666"
                        width="60"
                        x="5"
                        x1="59"
                        x2="65"
                        y="5"
                        y1="170"
                        y2="170"
                      />
                      <text
                        class="recharts-text recharts-cartesian-axis-tick-value"
                        fill="#666"
                        height="220"
                        orientation="left"
                        stroke="none"
                        text-anchor="end"
                        width="60"
                        x="57"
                        y="170"
                      >
                        <tspan
                          dy="0.355em"
                          x="57"
                        >
                          3
                        </tspan>
                      </text>
                    </g>
                    <g
                      class="recharts-layer recharts-cartesian-axis-tick"
                    >
                      <line
                        angle="0"
                        class="recharts-cartesian-axis-tick-line"
                        fill="none"
                        height="220"
                        orientation="left"
                        stroke="#666"
                        width="60"
                        x="5"
                        x1="59"
                        x2="65"
                        y="5"
                        y1="115"
                        y2="115"
                      />
                      <text
                        class="recharts-text recharts-cartesian-axis-tick-value"
                        fill="#666"
                        height="220"
                        orientation="left"
                        stroke="none"
                        text-anchor="end"
                        width="60"
                        x="57"
                        y="115"
                      >
                        <tspan
                          dy="0.355em"
                          x="57"
                        >
                          6
                        </tspan>
                      </text>
                    </g>
                    <g
                      class="recharts-layer recharts-cartesian-axis-tick"
                    >
                      <line
                        angle="0"
                        class="recharts-cartesian-axis-tick-line"
                        fill="none"
                        height="220"
                        orientation="left"
                        stroke="#666"
                        width="60"
                        x="5"
                        x1="59"
                        x2="65"
                        y="5"
                        y1="60"
                        y2="60"
                      />
                      <text
                        class="recharts-text recharts-cartesian-axis-tick-value"
                        fill="#666"
                        height="220"
                        orientation="left"
                        stroke="none"
                        text-anchor="end"
                        width="60"
                        x="57"
                        y="60"
                      >
                        <tspan
                          dy="0.355em"
                          x="57"
                        >
                          9
                        </tspan>
                      </text>
                    </g>
                    <g
                      class="recharts-layer recharts-cartesian-axis-tick"
                    >
                      <line
                        angle="0"
                        class="recharts-cartesian-axis-tick-line"
                        fill="none"
                        height="220"
                        orientation="left"
                        stroke="#666"
                        width="60"
                        x="5"
                        x1="59"
                        x2="65"
                        y="5"
                        y1="5"
                        y2="5"
                      />
                      <text
                        class="recharts-text recharts-cartesian-axis-tick-value"
                        fill="#666"
                        height="220"
                        orientation="left"
                        stroke="none"
                        text-anchor="end"
                        width="60"
                        x="57"
                        y="5"
                      >
                        <tspan
                          dy="0.355em"
                          x="57"
                        >
                          12
                        </tspan>
                      </text>
                    </g>
                  </g>
                </g>
                <g
                  class="recharts-layer recharts-bar"
                  id="recharts-bar-:r0:"
                >
                  <g
                    class="recharts-layer recharts-bar-rectangles"
                  >
                    <g
                      class="recharts-layer"
                    >
                      <g
                        class="recharts-layer recharts-bar-rectangle"
                      />
                      <g
                        class="recharts-layer recharts-bar-rectangle"
                      />
                      <g
                        class="recharts-layer recharts-bar-rectangle"
                      />
                    </g>
                  </g>
                </g>
              </svg>
            </div>
          </div>
        </div>
      </div>
      <table
        class="chart-card__table"
      >
        <caption
          class="visually-hidden"
        >
          Inventory overview data table
        </caption>
        <thead>
          <tr>
            <th
              scope="col"
            >
              Label
            </th>
            <th
              scope="col"
            >
              Properties
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              Available
            </td>
            <td>
              12
            </td>
          </tr>
          <tr>
            <td>
              Reserved
            </td>
            <td>
              5
            </td>
          </tr>
          <tr>
            <td>
              Sold
            </td>
            <td>
              7
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </section>
</DocumentFragment>
`;
